pkg/volume/glusterfs/glusterfs.go

```diff
@@ -436,6 +439,89 @@ func (d *glusterfsVolumeDeleter) GetPath() string {
 	return d.plugin.host.GetPodVolumeDir(d.glusterfsMounter.glusterfs.pod.UID, strings.EscapeQualifiedNameForDisk(name), d.glusterfsMounter.glusterfs.volName)
 }
 
+//
+// Traverse the PVs, fetching all the GIDs from those
+// in a given storage class, and mark them in the table.
+//
+func (p *glusterfsPlugin) collectGids(class *storage.StorageClass) error {
+	pvList, err := p.host.GetKubeClient().Core().PersistentVolumes().List(v1.ListOptions{LabelSelector: labels.Everything().String()})
+	if err != nil {
+		glog.Errorf("glusterfs: failed to get existing persistent volumes")
+		return err
+	}
+
+	var gidStrings []string
+	gidStrings = []
+
+	for _, pv := range pvList.Items {
+		if pv.Annotations[annDynamicallyProvisioned] != glusterfsPluginName {
+			continue
+		}
+
+		if pv.Annotations[annStorageClass] != class.name {
+			continue
+		}
+
+		gidStr := pv.Annotations[volumehelper.VolumeGidAnnotationKey]
+
+		gid, err := convertGid(gidStr)
+		if err != nil {
+			glog.Errorf("glusterfs: failed to convert gid %v", gidStr)
+			return err
+		}
+
+		err = p.gidTable[class.name].Allocate(gid)
+		if err != nil {
+			glog.Errorf("glusterfs: failed to store gid: %v", err)
+			return err
+		}
+	}
+
+	return nil
+}
+
+//
+// Return the gid table for a storage class.
+// - If this is the first time, fill it with all the gids
+//   used in PVs of this storage class by traversing the PVs.
+// - Adapt the range of the table to the current range of the SC.
+//
+func (p *glusterfsPlugin) getGidTable(class *storage.StorageClass) (*MinMaxAllocator, error) {
+	cfg, err := parseClassParameters(class.Parameters, p.host.GetKubeClient())
+	if err != nil {
+		return err
+	}
+
+	if p.gidTable[class.name] == nil {
```

Comment:
getGIDTable needs to be threadsafe, so guard gidTable with a lock, do a fast check under lock, build if needed, then do a follow up check under lock before returning

```
p.gidTableLock.Lock()
allocator, ok := p.gidTable[className]
p.gidTableLock.Unlock()
if ok {
  allocator.SetMinMax(min, max)
  return allocator, nil
}

// build, collect, set min/max ...
newAllocator := ... 

p.gidTableLock.Lock()
defer p.gidTableLock.Unlock()
// if one showed up since we last checked, use it
if allocator, ok := p.gidTable[className]; ok {
  allocator.SetMinMax(min, max)
  return allocator, nil
}
// otherwise, use the one we built
p.gidTable[className] = newAllocator
return newAllocator, nil
```